Wikipedia/Code/SectionEditorWebViewConfiguration.swift

```diff
@@ -0,0 +1,119 @@
+import WebKit
+
+extension NSNotification.Name {
+    static let WMFSectionEditorSelectionChangedNotification = Notification.Name("WMFSectionEditorSelectionChangedNotification")
+    static let WMFSectionEditorButtonHighlightNotification = Notification.Name("WMFSectionEditorButtonHighlightNotification")
+}
+
+extension SectionEditorWebViewConfiguration {
+    static let WMFSectionEditorSelectionChanged = "WMFSectionEditorSelectionChanged"
+    static let WMFSectionEditorSelectionChangedSelectedButton = "WMFSectionEditorSelectionChangedSelectedButton"
+}
+
+struct ButtonNeedsToBeSelectedMessage {
+    let type: EditButtonType
+    let ordered: Bool
+    let depth: Int
+}
+
+struct SelectionChangedMessage {
+    let selectionIsRange: Bool
+}
+
+private enum MessageNameConstants: String {
+    case selectionChanged
+    case highlightTheseButtons
+}
+
+private enum MessageConstants: String {
+    case button
+    case info
+}
+
+enum EditButtonType: String {
+    case li
+    case heading
+    case indent
+    case signature
+    case link
+    case bold
+    case italic
+    case reference
+    case template
+    case undo
+    case redo
+    case debug
+    case comment
+    case smallTextSize
+    case bigTextSize
+}
+
+private enum ButtonInfoConstants: String {
+    case ordered
+    case depth
+}
+
+class SectionEditorWebViewConfiguration: WKWebViewConfiguration, WKScriptMessageHandler {
+
+    override init() {
+        super.init()
+        setURLSchemeHandler(WMFURLSchemeHandler.shared(), forURLScheme: WMFURLSchemeHandlerScheme)
+        
+        let contentController = WKUserContentController()
+        contentController.add(self, name: MessageNameConstants.selectionChanged.rawValue)
+        contentController.add(self, name: MessageNameConstants.highlightTheseButtons.rawValue)
+        userContentController = contentController
```

Comment:
Rather than having the configuration object handle all of the script messages, could these be broken up into separate objects that route the business logic and update the associated views? There's not a clear flow of state or separation of concerns with these notifications being observed and responded to in several places.